File: jvm-experiments-commons/src/main/java/kr/kth/commons/timeperiod/YearAndQuarter.java
package kr.kth.commons.timeperiod;

import com.google.common.base.Objects;
import kr.kth.commons.base.Guard;
import kr.kth.commons.base.ValueObjectBase;
import lombok.Getter;
import lombok.Setter;

/**
 * 년도와 분기를 표현합니다.
 * User: sunghyouk.bae@gmail.com
 * Date: 12. 12. 25.
 */
public class YearAndQuarter extends ValueObjectBase implements Comparable<YearAndQuarter> {

	private static final long serialVersionUID = -9136294587041186630L;

	@Getter @Setter private Integer year;
	@Getter @Setter private QuarterKind quarter;

	public YearAndQuarter(Integer year, Integer quarter) {
		this(year, QuarterKind.valueOf(quarter));
	}

	public YearAndQuarter(Integer year, QuarterKind quarter) {
		this.year = year;
		this.quarter = quarter;
	}

	@Override
	public int compareTo(YearAndQuarter o) {
		Guard.shouldNotBeNull(o, "o");
		return this.hashCode() - o.hashCode();
	}

	@Override
	public int hashCode() {
		int yearNo = (year != null) ? year : 0;
		int quarterNo = (quarter != null) ? quarter.intValue() : 0;
		return yearNo * 100 + quarterNo * 25;
	}

	@Override
	protected Objects.ToStringHelper buildStringHelper() {
		return super.buildStringHelper()
		            .add("year", year)
		            .add("quarter", quarter);
	}
}

File: jvm-experiments-commons/src/main/scala/kr/kth/commons/timeperiod/YearAndQuarter.scala
package kr.kth.commons.timeperiod

import beans.BeanProperty
import kr.kth.commons.base.ValueObjectBase

/**
 * 년/분기 정보를 표현합니다.
 * User: sunghyouk.bae@gmail.com
 * Date: 12. 12. 26
 */
class YearAndQuarter(@BeanProperty var year: Int,
                     @BeanProperty var quarter: QuarterKind = QuarterKind.First)
    extends ValueObjectBase with Comparable[YearAndQuarter] {

    def this(year: Int, quarter: Int = 1) = this(year, QuarterKind.valueOf(quarter))

    def compareTo(other: YearAndQuarter): Int = hashCode().compareTo(other.hashCode())

    override def hashCode(): Int = year * 100 + (if (quarter != null) quarter.toInt else 0)

    protected override def buildStringHelper =
        super.buildStringHelper
            .add("year", year)
            .add("quarter", quarter)
}
